src/runtime/stream.rs

```diff
@@ -0,0 +1,293 @@
+
+    if socket_addrs.is_empty() {
+        return Err(ErrorKind::NoDnsResults(address.clone()).into());
+    }
+
+    // After considering various approaches, we decided to do what other drivers do, namely try each
+    // of the addresses in sequence with a preference for IPv4.
+    socket_addrs.sort_by_key(|addr| if addr.is_ipv4() { 0 } else { 1 });
+
+    let mut connect_error = None;
+
+    for address in &socket_addrs {
+        connect_error = match try_connect(address, timeout).await {
+            Ok(stream) => return Ok(stream),
+            Err(err) => Some(err),
+        };
+    }
+
+    Err(connect_error.unwrap_or_else(|| ErrorKind::NoDnsResults(address.clone()).into()))
+}
+
+impl AsyncStream {
+    /// Creates a new Tokio TCP stream connected to the server as specified by `options`.
+    #[cfg(feature = "tokio-runtime")]
+    pub(super) async fn connect_tokio(options: StreamOptions) -> Result<Self> {
+        use std::sync::Arc;
+
+        use tokio::net::TcpStream;
+        use tokio_rustls::TlsConnector;
+        use webpki::DNSNameRef;
+
+        let inner: TcpStream = connect_stream(&options.address, options.connect_timeout).await?;
+        inner.set_nodelay(true)?;
+
+        let inner = AsyncStreamInner::Tokio(inner);
+
+        // If there are TLS options, wrap the inner stream with rustls.
+        match options.tls_options {
+            Some(cfg) => {
+                let name = DNSNameRef::try_from_ascii_str(&options.address.hostname)?;
+                let mut tls_config = cfg.into_rustls_config()?;
+                tls_config.enable_sni = true;
+
+                let connector: TlsConnector = Arc::new(tls_config).into();
+                let session = connector.connect(name, inner).await?;
+
+                Ok(Self::Tls(session))
+            }
+            None => Ok(Self::Tcp(inner)),
+        }
+    }
+
+    #[cfg(feature = "async-std-runtime")]
+    pub(super) async fn connect_async_std(options: StreamOptions) -> Result<Self> {
+        use std::sync::Arc;
+
+        use async_std::net::TcpStream;
+        use tokio_rustls::TlsConnector;
+        use webpki::DNSNameRef;
+
+        let inner: TcpStream = connect_stream(&options.address, options.connect_timeout).await?;
+        inner.set_nodelay(true)?;
+
+        let inner = AsyncStreamInner::AsyncStd(inner);
+
+        // If there are TLS options, wrap the inner stream with rustls.
+        match options.tls_options {
+            Some(cfg) => {
+                let name = DNSNameRef::try_from_ascii_str(&options.address.hostname)?;
+                let mut tls_config = cfg.into_rustls_config()?;
+                tls_config.enable_sni = true;
+
+                let connector: TlsConnector = Arc::new(tls_config).into();
+                let session = connector.connect(name, inner).await?;
+
+                Ok(Self::Tls(session))
+            }
+            None => Ok(Self::Tcp(inner)),
+        }
+    }
+}
+
+impl AsyncRead for AsyncStream {
```

Comment:
The `AsyncRead` and `AsyncWrite` implementations on `AsyncStream` and `AsyncStreamInner` just delegate to the inner `TcpStream`, pinning where necessary.